zipline/pipeline/loaders/blaze/_core.pyx

```diff
@@ -0,0 +1,449 @@
+        )
+    elif kind == 'M':
+        _ffill_missing_value_2d_inplace_impl[np.int64_t](
+            array.view('int64'),
+            missing_value.astype('int64'),
+            is_missing_value[np.int64_t],
+        )
+    elif kind == 'f':
+        if isnan(missing_value):
+            _ffill_missing_value_2d_inplace_impl[np.float64_t](
+                array,
+                missing_value,
+                is_missing_nan,
+            )
+        else:
+            _ffill_missing_value_2d_inplace_impl[np.float64_t](
+                array,
+                missing_value,
+                is_missing_value[np.float64_t],
+            )
+    elif kind == 'O':
+        _ffill_missing_value_2d_inplace_impl[object](
+            array,
+            missing_value,
+            is_missing_value[object],
+        )
+    else:
+        raise TypeError('unknown column dtype: %r' % array.dtype)
+
+
+cdef _adjusted_array_for_column_impl(object dtype,
+                                     np.ndarray[column_type, ndim=2] out_array,
+                                     Py_ssize_t size,
+                                     np.ndarray[np.int64_t] ts_ixs,
+                                     np.ndarray[np.int64_t] asof_ixs,
+                                     np.ndarray[np.int64_t] sids,
+                                     dict column_ixs,
+                                     object mask,
+                                     np.ndarray[column_type] input_array,
+                                     column_type missing_value,
+                                     bint is_missing(column_type, column_type)):
+    cdef column_type value
+    cdef np.int64_t ts_ix
+    cdef np.int64_t asof_ix
+    cdef np.int64_t sid
+    cdef np.int64_t column_ix
+
+    cdef PyObject* adjustments_list_ptr
+    cdef list adjustments_list
+
+    cdef list non_null_ixs
+    cdef dict non_null_ixs_by_sid = {sid: [] for sid in sids}
+
+    cdef dict adjustments = {}
+
+    cdef set categories
+    if column_type is object:
+        # for object columns we need to maintain the unique values for the
+        # categories
+        categories = set()
+
+    cdef Py_ssize_t n
+    for n in range(size):
+        with cython.boundscheck(False), cython.wraparound(False):
+            value = input_array[n]
+
+        if is_missing(value, missing_value):
+            # skip missing values
+            continue
+
+        if column_type is object:
+            # maintain the categories for the label array if we have an object
+            # column
+            categories.add(value)
+
+        with cython.boundscheck(False), cython.wraparound(False):
+            ts_ix = ts_ixs[n]
+            asof_ix = asof_ixs[n]
+            sid = sids[n]
+
+        column_ix_ob = PyDict_GetItem(column_ixs, sid)
+        if column_ix_ob is NULL:
+            # ignore sids that are not requested
+            continue
+
+        column_ix = <object> column_ix_ob  # cast to np.int64_t
```

Comment:
`npy_intp` is `int*`, using that as an integral value is weird. `Py_ssize_t` and `npy_int64` are the same thing.